exercises/practice/two-bucket/.meta/example.ex

```diff
@@ -0,0 +1,55 @@
+defmodule TwoBucket do
+  defstruct [:goal, :moves, :other_bucket]
+  @type t :: %TwoBucket{goal: :one | :two, moves: integer, other_bucket: integer}
+
+  @doc """
+  Find the quickest way to fill a bucket with some amount of water from two buckets of specific sizes.
+  """
+  @spec measure(
+          size_one :: integer,
+          size_two :: integer,
+          goal :: integer,
+          start_bucket :: :one | :two
+        ) :: {:ok, TwoBucket.t()} | {:error, :impossible}
+  def measure(size_one, size_two, goal, start_bucket) do
+    other_filled = if start_bucket == :one, do: {0, size_two}, else: {size_one, 0}
+    forbidden_states = MapSet.new([other_filled])
+
+    # Partially apply function to avoid passing many arguments
+    next_moves = &next_moves(&1, size_one, size_two)
+
+    pour([{{0, 0}, 0}], forbidden_states, next_moves, goal)
+  end
+
+  def pour([], _, _, _), do: {:error, :impossible}
+
+  def pour([{{goal, other}, moves} | _], _, _, goal),
+    do: {:ok, %TwoBucket{goal: :one, moves: moves, other_bucket: other}}
+
+  def pour([{{other, goal}, moves} | _], _, _, goal),
+    do: {:ok, %TwoBucket{goal: :two, moves: moves, other_bucket: other}}
+
+  def pour([{state, moves} | states], forbidden_states, next_moves, goal) do
+    next =
+      next_moves.(state)
+      |> Enum.reject(&MapSet.member?(forbidden_states, &1))
+      |> Enum.map(&{&1, moves + 1})
+
+    pour(states ++ next, MapSet.put(forbidden_states, state), next_moves, goal)
```

Comment:
This is just a generic worry, I haven't analyzed the solution in detail, so it might be misguided, but:

Appending to a list in a loop? Are you sure this is the best way?

Reply:
Yes, because this is a breadth first search. I need to look at states with a lower number of moves first, and the next states I just created last. Otherwise I would need to generate all paths and check for the lowest after. 
A queue is the best data structure for this, but it's not worth implementing in this solution I think (and also the standard FP queue is made out of two lists and use that infamous `Enum.reverse` that you don't like :p ).

Reply:
All good. My comment about two lists and `Enum.reverse` was purely in the context of double linked lists 👻.

If you need a queue, use Erlang's (http://erlang.org/doc/man/queue.html). It has functions for adding and removing from both ends. And yes, it's two lists 😁